reflex-dom-core/src/Reflex/Dom/Attributes/Types.hs

```diff
@@ -0,0 +1,298 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | Types for creating or interacting with attributes.
+module Reflex.Dom.Attributes.Types
+  (
+  -- * Base types
+    Namespace
+  , AttributeName(..)
+
+  -- * Text based attribute functions
+  , declareAttribute
+  , setAttribute
+  , removeAttribute
+  , declareAttributeMap
+  , mapKeysToAttributeName
+
+  -- * Attributes
+  , DeclareAttrs
+  , attributesPatchIsEmpty
+  , removeAttrFromDeclareAttrs
+  , lookupAttrInPatch
+  , ModifyAttrs
+  , removeAttrFromModifyAttrs
+  , toPatch
+
+  -- * Implementing new attributes
+  , IsAttribute(..)
+  , PatchSettings(..)
+  , singleAttribute
+  , singleModifyAttribute
+  ) where
+
+import Control.Lens
+import Control.Monad (when, void)
+import Data.Dependent.Map (DMap)
+import Data.Dependent.Sum (DSum(..))
+import Data.Foldable (traverse_)
+import Data.GADT.Compare.TH (deriveGEq, deriveGCompare)
+import Data.Map (Map)
+import Data.Map.Misc (diffMap)
+import Data.Patch
+import Data.Proxy
+import Data.String (IsString(..))
+import Data.Text
+import GHCJS.DOM.Types (JSM)
+import Type.Reflection
+import qualified Data.Dependent.Map as DMap
+import qualified Data.Dependent.Map.Lens as DMap
+import qualified Data.Map as Map
+import qualified GHCJS.DOM.Element as Element
+import qualified GHCJS.DOM.Types as DOM
+
+type Namespace = Text
+
+-- | Attribute names which may or may not have namespaces
+data AttributeName = AttributeName !(Maybe Namespace) !Text deriving (Show, Read, Eq, Ord)
+
+-- | By default, AttributeNames are unnamespaced
+instance IsString AttributeName where
+  fromString = AttributeName Nothing . fromString
+
+-- | Used for controlling 'applyAttrPatchDOM'.
+data PatchSettings
```

Comment:
What does this type actually mean?  This seems like a very ad-hoc thing to impose on anyone who implements a custom attribute.

Reply:
It’s unfortunate for sure. But if we stop patching during hydration I think this can go away.